![logo-gt.png](attachment:logo-gt.png)

<div class="alert alert-block alert-info">
    <h1>BIOS 4150/BIOL 6150</h1>
    <h3>Instructor: Dr. King Jordan</h3>
    <p>Shivam Sharma (shivamsharma13@gatech.edu), Nilavrah Sensarma (nsensarma3@gatech.edu), Bengy Gyimah-Asamoah, (bgyimaha3@gatech.edu)

</p>
</div>

<div class="alert alert-block alert-warning">
    <h2>Project 8 (Pharmacogenomics) starter notebook</h2>
    <h3>Deadline: 11:59PM, November 18th, 2024</h3>
</div>

<div class="alert alert-block alert-danger">
    <h2>Alert!</h2>
    <p>Do not delete any comments or information boxes the starter notebook comes with</p>
</div>

---

# **1. Knowing your data**
### *Total Questions: 3*
### *Total Points: 5+5+20 = 30*


---

<div class="alert alert-block alert-warning">
    <h3>1.1 Download the information file</h3>
    <p>For this project, you will need to access PharmGKB annotated variants. PharmGKB has multiple resources, but we will focus on the clinical variant data file. This file should have variant-drug pairs and level of evidence for all clinical annotations.</p>
    <p>Download the file and show top 5 lines of the file</p>
</div>

In [1]:
#Show the file here.
!head -n 5  ~/biol6150/ProjectSubmissions/Group8/Project8/clinicalVariants.tsv

variant	gene	type	level of evidence	chemicals	phenotypes
CYP2C9*1, CYP2C9*2, CYP2C9*3, CYP2C9*13	CYP2C9	Metabolism/PK	1A	meloxicam	
CYP2C9*1, CYP2C9*3, CYP2C9*13	CYP2C9	Metabolism/PK	1A	lornoxicam	
CYP2C9*1, CYP2C9*2, CYP2C9*3	CYP2C9	Metabolism/PK	1A	siponimod	
rs17376848	DPYD	Toxicity	1A	capecitabine	Neoplasms


<div class="alert alert-block alert-warning">
    <h3>1.2 Filter the file</h3>
    <p>Although star alleles can be processed, we will only focus on dbSNP annotated variants (those that have rs ids)</p>
    <p>Filter the file for dbSNP annotated variants and count them</p>
    
</div>

In [ ]:
#The following code was run in R (can be found under 'Project8_Final.R' in the folder 'Rcode')
install.packages('readr')
library(readr)
data <- read_tsv("clinicalVariants.tsv")
filtered_data <- data[grep("^rs[0-9]+$", data$variant), ]
dim(filtered_data)
write.table(filtered_data, file = "filtered_clinicalVariants.tsv", sep = "\t", row.names = FALSE, col.names = TRUE, quote = FALSE)
#The output for dim(filtered_data) is Rows: 4575 Columns: 6

In [2]:
# wc -l after filtering 
!wc -l ~/biol6150/ProjectSubmissions/Group8/Project8/filtered_clinicalVariants.tsv

4576 /home/hice1/acressy3/biol6150/ProjectSubmissions/Group8/Project8/filtered_clinicalVariants.tsv


4576 includes the header which means there are 4575 dbSNP annotated variants

<div class="alert alert-block alert-warning">
    <h3>1.3 Think ahead!</h3>
    <p><b>Eventual Goal:</b> Quantify the presence of the variants obtained in Question 1.2. We will extract these variants from the 1000 genomes VCF file.</p>
    <p>Use whatever resource you want to. The end goal is to have appropriate information about these variants obtained in 1.2.</p>
    <p>If you have to subset a VCF file (extract variants of interest), you will need information such as CHR, POS, etc. Although you can simply use the ID column to extract variants, that might not work everywhere.</p>
    <p>If you want to use R, please feel free to do so. Add screenshots from PACE-ICE RStudio here. Your R workflow would essentially output a nice tsv file which you use later to extract variants</p>
    
</div>

<div class="alert alert-block alert-danger">
    <h2>Alert!</h2>
    <p>We do not want you to use dbSNP IDs (rs ids) to extract variants from the 1000 genomes data. The 1000 genomes file we provided you has dbSNP IDs and a simple bcftools command will do the job. But most VCF files you will come across in your work will not have dbSNP IDs. Therefore, we expect you to find relevant information about the variants obtained in Question 1.2 before moving to section 2.</p>
</div>

In [ ]:
#The end goal is to have variants 

In [ ]:
#After speaking with Shivam, we are moving forward using the rsIDs

# **2. Extract data**
### *Total Questions: 2*
### *Total Points: 5+15=20*


---

<div class="alert alert-block alert-warning">
    <h3>2.1 Subset the VCF files</h3>
    <p>2.1.1 How many variants obtained in 1.2 were found in the 1000 genomes data?</p>
    <p>2.1.2 Subset the VCF files (all chromosomes) and write the number of Pharmacogenomic variants found in chromosome 2.</p>
</div>

In [2]:
#All commands/code goes here.

In [1]:
#Show python, R or bash code here.
#Extract clinical variants from the 1000 genomes chromosomal data
!bcftools view -i 'ID=@clinicalVariants_rsIDs.txt' ~/biol6150/Data/1000Genomes/phase3.chr1.GRCh38.GT.crossmap.vcf.gz > ~/biol6150/ProjectSubmissions/Group8/Project8/clinvar_1000G/clinvar_1000G_chr1.vcf
print("chr 1 done")
!bcftools view -i 'ID=@clinicalVariants_rsIDs.txt' ~/biol6150/Data/1000Genomes/phase3.chr2.GRCh38.GT.crossmap.vcf.gz > ~/biol6150/ProjectSubmissions/Group8/Project8/clinvar_1000G/clinvar_1000G_chr2.vcf
print("chr 2 done")
!bcftools view -i 'ID=@clinicalVariants_rsIDs.txt' ~/biol6150/Data/1000Genomes/phase3.chr3.GRCh38.GT.crossmap.vcf.gz > ~/biol6150/ProjectSubmissions/Group8/Project8/clinvar_1000G/clinvar_1000G_chr3.vcf
print("chr 3 done")
!bcftools view -i 'ID=@clinicalVariants_rsIDs.txt' ~/biol6150/Data/1000Genomes/phase3.chr4.GRCh38.GT.crossmap.vcf.gz > ~/biol6150/ProjectSubmissions/Group8/Project8/clinvar_1000G/clinvar_1000G_chr4.vcf
print("chr 4 done")
!bcftools view -i 'ID=@clinicalVariants_rsIDs.txt' ~/biol6150/Data/1000Genomes/phase3.chr5.GRCh38.GT.crossmap.vcf.gz > ~/biol6150/ProjectSubmissions/Group8/Project8/clinvar_1000G/clinvar_1000G_chr5.vcf
print("chr 5 done")
!bcftools view -i 'ID=@clinicalVariants_rsIDs.txt' ~/biol6150/Data/1000Genomes/phase3.chr6.GRCh38.GT.crossmap.vcf.gz > ~/biol6150/ProjectSubmissions/Group8/Project8/clinvar_1000G/clinvar_1000G_chr6.vcf
print("chr 6 done")
!bcftools view -i 'ID=@clinicalVariants_rsIDs.txt' ~/biol6150/Data/1000Genomes/phase3.chr7.GRCh38.GT.crossmap.vcf.gz > ~/biol6150/ProjectSubmissions/Group8/Project8/clinvar_1000G/clinvar_1000G_chr7.vcf
print("chr 7 done")
!bcftools view -i 'ID=@clinicalVariants_rsIDs.txt' ~/biol6150/Data/1000Genomes/phase3.chr8.GRCh38.GT.crossmap.vcf.gz > ~/biol6150/ProjectSubmissions/Group8/Project8/clinvar_1000G/clinvar_1000G_chr8.vcf
print("chr 8 done")
!bcftools view -i 'ID=@clinicalVariants_rsIDs.txt' ~/biol6150/Data/1000Genomes/phase3.chr9.GRCh38.GT.crossmap.vcf.gz > ~/biol6150/ProjectSubmissions/Group8/Project8/clinvar_1000G/clinvar_1000G_chr9.vcf
print("chr 9 done")
!bcftools view -i 'ID=@clinicalVariants_rsIDs.txt' ~/biol6150/Data/1000Genomes/phase3.chr10.GRCh38.GT.crossmap.vcf.gz > ~/biol6150/ProjectSubmissions/Group8/Project8/clinvar_1000G/clinvar_1000G_chr10.vcf
print("chr 10 done")
!bcftools view -i 'ID=@clinicalVariants_rsIDs.txt' ~/biol6150/Data/1000Genomes/phase3.chr11.GRCh38.GT.crossmap.vcf.gz > ~/biol6150/ProjectSubmissions/Group8/Project8/clinvar_1000G/clinvar_1000G_chr11.vcf
print("chr 11 done")
!bcftools view -i 'ID=@clinicalVariants_rsIDs.txt' ~/biol6150/Data/1000Genomes/phase3.chr12.GRCh38.GT.crossmap.vcf.gz > ~/biol6150/ProjectSubmissions/Group8/Project8/clinvar_1000G/clinvar_1000G_chr12.vcf
print("chr 12 done")
!bcftools view -i 'ID=@clinicalVariants_rsIDs.txt' ~/biol6150/Data/1000Genomes/phase3.chr13.GRCh38.GT.crossmap.vcf.gz > ~/biol6150/ProjectSubmissions/Group8/Project8/clinvar_1000G/clinvar_1000G_chr13.vcf
print("chr 13 done")
!bcftools view -i 'ID=@clinicalVariants_rsIDs.txt' ~/biol6150/Data/1000Genomes/phase3.chr14.GRCh38.GT.crossmap.vcf.gz > ~/biol6150/ProjectSubmissions/Group8/Project8/clinvar_1000G/clinvar_1000G_chr14.vcf
print("chr 14 done")
!bcftools view -i 'ID=@clinicalVariants_rsIDs.txt' ~/biol6150/Data/1000Genomes/phase3.chr15.GRCh38.GT.crossmap.vcf.gz > ~/biol6150/ProjectSubmissions/Group8/Project8/clinvar_1000G/clinvar_1000G_chr15.vcf
print("chr 15 done")
!bcftools view -i 'ID=@clinicalVariants_rsIDs.txt' ~/biol6150/Data/1000Genomes/phase3.chr16.GRCh38.GT.crossmap.vcf.gz > ~/biol6150/ProjectSubmissions/Group8/Project8/clinvar_1000G/clinvar_1000G_chr16.vcf
print("chr 16 done")
!bcftools view -i 'ID=@clinicalVariants_rsIDs.txt' ~/biol6150/Data/1000Genomes/phase3.chr17.GRCh38.GT.crossmap.vcf.gz > ~/biol6150/ProjectSubmissions/Group8/Project8/clinvar_1000G/clinvar_1000G_chr17.vcf
print("chr 17 done")
!bcftools view -i 'ID=@clinicalVariants_rsIDs.txt' ~/biol6150/Data/1000Genomes/phase3.chr18.GRCh38.GT.crossmap.vcf.gz > ~/biol6150/ProjectSubmissions/Group8/Project8/clinvar_1000G/clinvar_1000G_chr18.vcf
print("chr 18 done")
!bcftools view -i 'ID=@clinicalVariants_rsIDs.txt' ~/biol6150/Data/1000Genomes/phase3.chr19.GRCh38.GT.crossmap.vcf.gz > ~/biol6150/ProjectSubmissions/Group8/Project8/clinvar_1000G/clinvar_1000G_chr19.vcf
print("chr 19 done")
!bcftools view -i 'ID=@clinicalVariants_rsIDs.txt' ~/biol6150/Data/1000Genomes/phase3.chr20.GRCh38.GT.crossmap.vcf.gz > ~/biol6150/ProjectSubmissions/Group8/Project8/clinvar_1000G/clinvar_1000G_chr20.vcf
print("chr 20 done")
!bcftools view -i 'ID=@clinicalVariants_rsIDs.txt' ~/biol6150/Data/1000Genomes/phase3.chr21.GRCh38.GT.crossmap.vcf.gz > ~/biol6150/ProjectSubmissions/Group8/Project8/clinvar_1000G/clinvar_1000G_chr21.vcf
print("chr 21 done")
!bcftools view -i 'ID=@clinicalVariants_rsIDs.txt' ~/biol6150/Data/1000Genomes/phase3.chr22.GRCh38.GT.crossmap.vcf.gz > ~/biol6150/ProjectSubmissions/Group8/Project8/clinvar_1000G/clinvar_1000G_chr22.vcf
print("chr 22 done")

chr 1 done
chr 2 done
chr 3 done
chr 4 done
chr 5 done
chr 6 done
chr 7 done
chr 8 done
chr 9 done
chr 10 done
chr 11 done
chr 12 done
chr 13 done
chr 14 done
chr 15 done
chr 16 done
chr 17 done
chr 18 done
chr 19 done
chr 20 done
chr 21 done
chr 22 done


In [1]:
#combine all of the individual chromosome .vcf files into one .vcf file
!bcftools concat -o ~/biol6150/ProjectSubmissions/Group8/Project8/clinvar_1000G_all.vcf -O v ~/biol6150/ProjectSubmissions/Group8/Project8/clinvar_1000G/clinvar_1000G_chr*.vcf

Checking the headers and starting positions of 22 files
Concatenating /home/hice1/acressy3/biol6150/ProjectSubmissions/Group8/Project8/clinvar_1000G/clinvar_1000G_chr10.vcf	0.005921 seconds
Concatenating /home/hice1/acressy3/biol6150/ProjectSubmissions/Group8/Project8/clinvar_1000G/clinvar_1000G_chr11.vcf	0.004811 seconds
Concatenating /home/hice1/acressy3/biol6150/ProjectSubmissions/Group8/Project8/clinvar_1000G/clinvar_1000G_chr12.vcf	0.004859 seconds
Concatenating /home/hice1/acressy3/biol6150/ProjectSubmissions/Group8/Project8/clinvar_1000G/clinvar_1000G_chr13.vcf	0.004263 seconds
Concatenating /home/hice1/acressy3/biol6150/ProjectSubmissions/Group8/Project8/clinvar_1000G/clinvar_1000G_chr14.vcf	0.004388 seconds
Concatenating /home/hice1/acressy3/biol6150/ProjectSubmissions/Group8/Project8/clinvar_1000G/clinvar_1000G_chr15.vcf	0.004561 seconds
Concatenating /home/hice1/acressy3/biol6150/ProjectSubmissions/Group8/Project8/clinvar_1000G/clinvar_1000G_chr16.vcf	0.004880 seconds
Concat

In [1]:
#check number of clinical variants found in 1000 genomes data
!bcftools view -H ~/biol6150/ProjectSubmissions/Group8/Project8/clinvar_1000G_all.vcf | wc -l

2542


In [3]:
#check number pharmocogenomic variants found on chromosome 2
!bcftools view -H ~/biol6150/ProjectSubmissions/Group8/Project8/clinvar_1000G/clinvar_1000G_chr2.vcf | wc -l

179


2.1.1 2542 variants from 1.2 were found in the 1000 genomes data.

2.1.2 There are 179 pharmacogenomic variants found on chromosome 2

# **3. Pharmacogenomic variant distribution in 1000 genomes populations**
### *Total Steps: 3*
### *Total Points:  25+5+20=50*


---

<div class="alert alert-block alert-warning">
    <h3>3.1 Distribution by populations</h3>
        <p>For the 26 populations present in the 1000 genomes data, show the distribution of pharmaocgenomic variant frequencies stratified by "level of evidence"</p><hr>
    <h4>Step 1 - Find variant frequencies for all pharmacogenomic variants within each population.</h4>
    <p>The end result of this step should be a TSV file which has <i>x rows and 26 + w columns</i>. x = number of pharmacogenomic variants found in 1000 genome VCF files. 26 columns represent each 1000 genome population. w will be additonal metadata about the pharmacogenomic variant such as CHR POS ID etc</p>
    <p>This is a good exercise to implement a nice bash or python based solution where you can streamline a lot of things.</p>
    <h4>Step 2 - Annotate variants by level of evidence</h4>
    <p>This should be easy, just annotate them in your code or create separate files.</p>
    <h4>Step 3 - Visualize using boxplots</h4>
    <p>Every level of evidence will have it's set of 26 box plots showing frequency distribution of all variants </p>
    <p>In the end you will have multiple box plot graphs like the one shown below. Each graph will correspond to a specific level of evidence as determined by PharmaGKB (1A, 1B, 2A, 2B, ... , etc)</p>
</div>

![Screenshot%202024-11-06%20at%2011.55.00%20PM-2.png](attachment:Screenshot%202024-11-06%20at%2011.55.00%20PM-2.png)

In [3]:
#All work goes here.

In [ ]:
#Step 1 - Find variant frequencies for all pharmacogenomic variants within each population.

In [1]:
#download the population info file
!wget https://ftp.1000genomes.ebi.ac.uk/vol1/ftp/release/20130502/integrated_call_samples_v3.20130502.ALL.panel

--2024-11-17 14:40:59--  https://ftp.1000genomes.ebi.ac.uk/vol1/ftp/release/20130502/integrated_call_samples_v3.20130502.ALL.panel
Resolving ftp.1000genomes.ebi.ac.uk (ftp.1000genomes.ebi.ac.uk)... 193.62.193.167
Connecting to ftp.1000genomes.ebi.ac.uk (ftp.1000genomes.ebi.ac.uk)|193.62.193.167|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: 55156 (54K)
Saving to: ‘integrated_call_samples_v3.20130502.ALL.panel.1’

integrated_call_sam 100%[===================>]  53.86K   191KB/s    in 0.3s    

2024-11-17 14:41:00 (191 KB/s) - ‘integrated_call_samples_v3.20130502.ALL.panel.1’ saved [55156/55156]



In [1]:
#Associate individuals with their variants (based on chromosome position not rsIDs)
!bcftools query -i 'GT~"1"' -f '%CHROM\t%POS\t[%SAMPLE\n]' ~/biol6150/ProjectSubmissions/Group8/Project8/clinvar_1000G_all.vcf > ~/biol6150/ProjectSubmissions/Group8/Project8/individuals_with_variants.txt

In [2]:
#Associate individuals with their populations
!sort -k1,1 ~/biol6150/ProjectSubmissions/Group8/Project8/integrated_call_samples_v3.20130502.ALL.panel > ~/biol6150/ProjectSubmissions/Group8/Project8/sorted_panel.txt
!sort -k3,3 ~/biol6150/ProjectSubmissions/Group8/Project8/individuals_with_variants.txt > ~/biol6150/ProjectSubmissions/Group8/Project8/sorted_individuals.txt
!join -1 3 -2 1 ~/biol6150/ProjectSubmissions/Group8/Project8/sorted_individuals.txt ~/biol6150/ProjectSubmissions/Group8/Project8/sorted_panel.txt > ~/biol6150/ProjectSubmissions/Group8/Project8/individuals_with_population.txt

In [1]:
#converting the 1000 genomes clinical variants .vcf file into a .tsv file
!bcftools query -f '%CHROM\t%POS\t%ID\t%REF\t%ALT[\t%SAMPLE=%GT]\n' ~/biol6150/ProjectSubmissions/Group8/Project8/clinvar_1000G_all.vcf > ~/biol6150/ProjectSubmissions/Group8/Project8/clinvar_1000G.tsv

In [1]:
#making a .tsv file containing sampleID and population
!cut -f 1,2 ~/biol6150/ProjectSubmissions/Group8/Project8/integrated_call_samples_v3.20130502.ALL.panel > ~/biol6150/ProjectSubmissions/Group8/Project8/sample_to_population.tsv

In [1]:
import pandas as pd

# Assign column names manually since the .tsv file doesn't have any
expected_columns = ["CHR", "POS", "ID", "REF", "ALT"]

# Extend the list with generic sample-genotype names
with open("clinvar_1000G.tsv") as file:
    sample_columns = [f"Sample_{i}" for i in range(
        len(file.readline().strip().split("\t")) - len(expected_columns)
    )]
columns = expected_columns + sample_columns

# Load the file with specified column names
variants = pd.read_csv("clinvar_1000G.tsv", sep="\t", header=None, names=columns)

# Reshape the DataFrame to long format
variants_long = variants.melt(
    id_vars=["CHR", "POS", "ID", "REF", "ALT"],  # Metadata columns
    var_name="Sample_GT",  # New column for sample-genotype pairs
    value_name="Genotype"  # Genotype values
)

# Split the 'Genotype' column into 'Sample' and 'GT'
variants_long[["Sample", "GT"]] = variants_long["Genotype"].str.split("=", expand=True)

# Drop the now unnecessary columns for clarity
variants_long = variants_long.drop(columns=["Sample_GT", "Genotype"])

# Verify the result
print(variants_long.head())

# Load the panel file which links sample IDs to population names
panel = pd.read_csv("sample_to_population.tsv", sep="\t", names=["Sample_ID", "Population"])

# Merge the population data with the genotype data
merged = variants_long.merge(panel, left_on="Sample", right_on="Sample_ID", how="left")

# Check merged result
print(merged.head())

# To make the kernel less stressed out
# Delete large variables that aren't being used anymore
del variants

# **Calculate variant frequency instead of genotype frequency**

# Create a column indicating the presence of the variant (i.e., GT is not 0|0)
merged['Variant_Present'] = merged['GT'].apply(lambda x: 1 if x != '0|0' else 0)

# Count the number of samples with each variant (variant present = 1)
variant_counts = merged.groupby(["CHR", "POS", "ID", "REF", "ALT", "Population"])["Variant_Present"].sum().reset_index(name="Variant_Count")

# Calculate the total number of individuals in each population
total_samples = merged.groupby(["Population"])["Sample"].nunique().reset_index(name="Total_Samples")

# Merge the total sample count into the variant counts
variant_counts = variant_counts.merge(total_samples, on="Population")

# Calculate the frequency of each variant
variant_counts["Frequency"] = variant_counts["Variant_Count"] / variant_counts["Total_Samples"]

# Check the resulting frequencies
print(variant_counts.head())

# Pivot the data to have population columns
final_variant_data = variant_counts.pivot_table(
    index=["CHR", "POS", "ID", "REF", "ALT"],
    columns="Population",
    values="Frequency",
    aggfunc="first"
).reset_index()

# Check the final structure
print(final_variant_data.head())

# Save the final result to a TSV file
final_variant_data.to_csv("final_variant_frequencies.tsv", sep="\t", index=False)

     CHR      POS          ID REF ALT   Sample   GT
0  chr10  4198678  rs11252394   G   A  HG00096  1|0
1  chr10  4768369   rs1901633   A   G  HG00096  1|0
2  chr10  5099115   rs1937840   C   G  HG00096  1|0
3  chr10  5102576  rs62621365   C   T  HG00096  0|0
4  chr10  5205791  rs11253043   G   A  HG00096  0|0
     CHR      POS          ID REF ALT   Sample   GT Sample_ID Population
0  chr10  4198678  rs11252394   G   A  HG00096  1|0   HG00096        GBR
1  chr10  4768369   rs1901633   A   G  HG00096  1|0   HG00096        GBR
2  chr10  5099115   rs1937840   C   G  HG00096  1|0   HG00096        GBR
3  chr10  5102576  rs62621365   C   T  HG00096  0|0   HG00096        GBR
4  chr10  5205791  rs11253043   G   A  HG00096  0|0   HG00096        GBR
    CHR       POS         ID REF ALT Population  Variant_Count  Total_Samples  \
0  chr1  11238733  rs2024627   T   C        ACB             56             96   
1  chr1  11792243  rs1476413   C   T        ACB             31             96   
2  chr1

In [1]:
#Make sure that the number of rows = number of clinical variants found in 1000 genomes
!awk 'NR > 1' ~/biol6150/ProjectSubmissions/Group8/Project8/final_variant_frequencies.tsv | wc -l

2542


In [ ]:
#Step 2 - Annotate variants by level of evidence

In [ ]:
#The following was used in R to add the levels of evidence to each variant
#All R code 'Project8_Final.R' in the folder 'Rcode'
#Output files are also found in the folder 'Rcode'

# Extract the rsid (1st column) and level of evidence (4th column)
# filtered_data was defined in Part 1 "Knowing your data"
level_of_evidence_data <- filtered_data[, c(1, 4)]
colnames(level_of_evidence_data) <- c("ID", "Level_of_Evidence")
# Save the level of evidence data to a new file
write.table(level_of_evidence_data, "variant_levels_annotated.tsv", sep = "\t", row.names = FALSE, col.names = TRUE)
# Read the file we just created in Step 2
final_fequency <-  fread("final_variant_frequencies.tsv", sep = "\t")
# Remove duplicates fro the level of evidence data and merge the .tsv files to create the annotated variants .tsv file
level_of_evidence_data_unique <- level_of_evidence_data[!duplicated(level_of_evidence_data$ID), ]
merged_data <- merge(final_fequency, level_of_evidence_data_unique, by = "ID", all.x = TRUE)
write.table(merged_data, "annotated_variant_frequencies.tsv", sep = "\t", row.names = FALSE, col.names = TRUE)

In [3]:
#Showing the first 3 lines of the file
!head -n 3 ~/biol6150/ProjectSubmissions/Group8/Project8/Rcode/annotated_variant_frequencies.tsv

"ID"	"CHR"	"POS"	"REF"	"ALT"	"ACB"	"ASW"	"BEB"	"CDX"	"CEU"	"CHB"	"CHS"	"CLM"	"ESN"	"FIN"	"GBR"	"GIH"	"GWD"	"IBS"	"ITU"	"JPT"	"KHV"	"LWK"	"MSL"	"MXL"	"PEL"	"PJL"	"PUR"	"STU"	"TSI"	"YRI"	"Level_of_Evidence"
"rs10007051"	"chr4"	129244309	"C"	"T"	0.59375	0.557377049180328	0.63953488372093	0.795698924731183	0.505050505050505	0.660194174757282	0.733333333333333	0.382978723404255	0.636363636363636	0.555555555555556	0.516483516483517	0.543689320388349	0.548672566371681	0.448598130841121	0.627450980392157	0.701923076923077	0.787878787878788	0.555555555555556	0.647058823529412	0.296875	0.164705882352941	0.635416666666667	0.471153846153846	0.53921568627451	0.495327102803738	0.592592592592593	"3"
"rs1000940"	"chr17"	5379957	"A"	"G"	0.354166666666667	0.409836065573771	0.697674418604651	0.903225806451613	0.474747474747475	0.796116504854369	0.885714285714286	0.617021276595745	0.313131313131313	0.737373737373737	0.626373626373626	0.621359223300971	0.52212389380531	0.532710280373832	0.696078431372549	0

In [1]:
#Make sure that the number of rows = number of clinical variants found in 1000 genomes
!awk 'NR > 1' ~/biol6150/ProjectSubmissions/Group8/Project8/Rcode/annotated_variant_frequencies.tsv | wc -l

2542


In [ ]:
#Step 3 - Visualize using boxplots

In [ ]:
##### Step 3 - Visualize using boxplots ####
#All R code 'Project8_Final.R' in the folder 'Rcode'
#Output files are also found in the folder 'Rcode'

## Assign superpopulations to the populations
# Define the list of population codes
populations <- c("ACB", "ASW", "BEB", "CDX", "CEU", "CHB", "CHS", "CLM", "ESN", 
                 "FIN", "GBR", "GIH", "GWD", "IBS", "ITU", "JPT", "KHV", "LWK", 
                 "MSL", "MXL", "PEL", "PJL", "PUR", "STU", "TSI", "YRI")
# Define the superpopulations associations
superpopulations <- c(
  "ACB" = "AFR", "ASW" = "AFR", "BEB" = "SAS", "CDX" = "EAS", 
  "CEU" = "EUR", "CHB" = "EAS", "CHS" = "EAS", "CLM" = "AMR", 
  "ESN" = "AFR", "FIN" = "EUR", "GBR" = "EUR", "GIH" = "SAS", 
  "GWD" = "AFR", "IBS" = "EUR", "ITU" = "EAS", "JPT" = "EAS", 
  "KHV" = "EAS", "LWK" = "AFR", "MSL" = "AFR", "MXL" = "AMR", 
  "PEL" = "AMR", "PJL" = "SAS", "PUR" = "AMR", "STU" = "SAS", 
  "TSI" = "EUR", "YRI" = "AFR"
)
# Create a vector of superpopulations based on the population codes
pop_to_super <- superpopulations[populations]
# Convert to a data frame for better readability
pop_super_df <- data.frame(Population = populations, Superpopulation = pop_to_super)

## Create a new table that adds a superpopulation based on the population and turns populations into a variable rather than a column
#This should make boxplots easier to make
merged_data_long <- merged_data %>%
  pivot_longer(cols = all_of(populations), 
               names_to = "Population", 
               values_to = "Genotype") %>%
  mutate(Superpopulation = superpopulations[Population])

# Ensure 'Population' is ordered by 'Superpopulation'
# This makes the graph look more clean and organized in the end
merged_data_long$Population <- factor(merged_data_long$Population, 
                                      levels = unique(merged_data_long$Population[order(merged_data_long$Superpopulation)]))

# Check what all of the levels of evidence we have are
unique(merged_data_long$Level_of_Evidence)
# Create a levels of evidence factor to help create separate plots
merged_data_long$Level_of_Evidence <- factor(merged_data_long$Level_of_Evidence, 
                                             levels = c("1A", "1B", "2A", "2B", "3", "4"))

# Loop over each level of evidence to create a separate boxplot
for(level in levels(merged_data_long$Level_of_Evidence)) {
  # Filter the data for the current level of evidence
  data_subset <- merged_data_long %>% filter(Level_of_Evidence == level)
  
  # Create and save the boxplots
  p <- ggplot(data_subset, aes(x = Population, y = Genotype, fill = Superpopulation)) + 
    geom_boxplot() + 
    scale_fill_manual(values = c("AFR" = "lightblue", "EUR" = "pink", "SAS" = "lightgreen", "EAS" = "mediumpurple1", "AMR" = "lightsalmon")) +  # Assign colors
    theme_minimal() + 
    theme(axis.text.x = element_text(angle = 90, hjust = 1)) +  # Rotate x-axis labels for readability
    labs(title = paste("Variant Frequency Distribution for Level of Evidence", level), # Create labels
         x = "Population", 
         y = "Variant Frequency",
         fill = "Superpopulation")
  
  # Print and save the plot
  print(p)
  ggsave(paste0("Boxplot_Level_", level, ".png"), plot = p, width = 10, height = 6)
}

#Plots are listed below

![Boxplot_Level_1A.png](attachment:Boxplot_Level_1A.png)

![Boxplot_Level_1B.png](attachment:Boxplot_Level_1B.png)

![Boxplot_Level_2A.png](attachment:Boxplot_Level_2A.png)

![Boxplot_Level_2B.png](attachment:Boxplot_Level_2B.png)

![Boxplot_Level_3.png](attachment:Boxplot_Level_3.png)

![Boxplot_Level_4.png](attachment:Boxplot_Level_4.png)

We have used the suggested method of inserting pictures and apologize if you cannot see them in the notebook. All pictures are also listed in the "BOXPLOTS" folder within our Project8 folder.

<div class="alert alert-block alert-danger">
    <h2>Alert!</h2>
    <p>If you use R, please provide ample evidence of work.</p>
    <p>For R use, you should add appropriate screenshots and the R code here in this notebook as well. R code might not run for you on PACE-ICE Jupyter engine right now, but it will work for anyone replicating this work somehwere else (that anyone could be the "future you")</p>
</div>